.github/workflows/main.yml

```diff
@@ -0,0 +1,236 @@
+      - name: Download spec tests
+        if: matrix.os != 'windows-latest'
+        run: yarn download-spec-tests
+
+        # Cache Spec Tests if not already cached
+      - name: Save spec tests to cache
+        if: matrix.os != 'windows-latest' && steps.cache-restore.outputs.cache-hit != 'true'
+        uses: actions/cache/save@v4
+        with:
+          path: spec-tests
+          key: spec-test-data-${{ hashFiles('test/spec/specTestVersioning.ts') }}
+
+      - name: Run spec tests
+        if: matrix.os != 'windows-latest'
+        run: yarn test:spec
+
+      - name: Upload binding.node
+        uses: actions/upload-artifact@v4
+        if: github.repository_owner == 'chainsafe' && github.event_name != 'pull_request'
+        with:
+          name: binding.node
+          path: prebuild/*.node
+          if-no-files-found: error
+
+  # Required as current base images do not support arm64
+  # https://resources.github.com/devops/accelerate-your-cicd-with-arm-and-gpu-runners-in-github-actions/
+  # https://github.com/actions/runner-images/issues/5631
+  build-arm:
+    strategy:
+      fail-fast: false
+      matrix:
+        node:
+          - 16
+          - 18
+          - 20
+          - 21
+    runs-on: ubuntu-22.04
+    defaults:
+      run:
+        shell: bash
+
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          submodules: true
+
+      # Download spec tests with cache
+      - name: Restore spec tests cache
+        uses: actions/cache@v4
+        with:
+          path: spec-tests
+          key: spec-test-data-${{ hashFiles('test/spec/specTestVersioning.ts') }}
+
+      # https://github.com/docker/setup-qemu-action
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@v3
+
+      # https://github.com/docker/setup-buildx-action
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v3
+
+      - name: Build arm64 Binding, Run Tests and Copy Build Artifact to Host
+        run: |
+          docker buildx build --build-arg NODE_VERSION=${{matrix.node}} --platform linux/arm64 -t blst-ts-test-armbuild:${{matrix.node}} --load --progress=plain .
+          docker create --name temp-container blst-ts-test-armbuild:${{matrix.node}}
+          docker cp temp-container:/usr/src/blst-ts/prebuild ./prebuild
+
+      - name: Upload binding.node
+        uses: actions/upload-artifact@v4
+        if: github.repository_owner == 'chainsafe' && github.event_name != 'pull_request'
+        with:
+          name: binding.node
+          path: prebuild/*.node
+          if-no-files-found: error
+
+  benchmark:
+    runs-on: ubuntu-22.04
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          submodules: true
+
+      - uses: actions/setup-node@v4
+        with:
+          node-version: 20
+
+      - name: Install
+        run: yarn
+
+      - name: Run Perf Tests
+        run: yarn test:perf
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+
+  publish:
```

Comment:
when do we trigger this? consider making it a separate workflow to either trigger it manually or whenever we create a tag
```yml
on:
  push:
    tags:
      - v*
```

Reply:
Good point to explain this further.  I added `CONTRIBUTING.md` that explains the publish workflow.  The way it works now is all jobs, other than `publish` run on `pull_request` and then when the PR gets merged to master the `push` trigger fires.  Its only setup to listen for pushes to master which can only happen via CI because of the branch protection.

When the push happens the ref will === `refs/heads/master` which will trigger the publish workflow.  There is an action in there that checks for changes to the package.json version key and if a new version makes it to master then it automagically creates a repo tag for the package.json version.  The following steps check to see if a tag was created and if so the workflow publishes the release and npm package for the new tag.

No more manually tagging necessary.